compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5326,6 +5311,56 @@ private void visitCheckAndCheckPanicExpr(BLangCheckedExpr checkedExpr) {
         resultType = types.checkType(checkedExpr, actualType, expType);
     }
 
+    private void rewriteWithEnsureTypeFunc(BLangCheckedExpr checkedExpr, BLangExpression expr,
+                                           BType typeOfExprWithCheckingKeyword) {
+        BType rhsType = getCandidateType(checkedExpr);
+        if (rhsType == symTable.semanticError) {
+            rhsType = checkExpr(checkedExpr.expr, env);
+        }
+        BType errorLiftedType = types.getSafeType(rhsType, false, true);
```

Comment:
`Types.getSafeType` doesn't seem to handle all cases. Right now, for error lifting it only removes the [builtin error](https://github.com/ballerina-platform/ballerina-lang/blob/e9d7c054e4561ba41890e00bbe5240552f60664d/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L4847). So the following doesn't work. Instead we should check by tag and remove.

```ballerina
public function main() returns Error? {
    () x = check func();
}

function func() returns json|Error {

}

type Error distinct error;
```

Also, although probably has no impact on this path, we don't seem to be checking if nil should be lifted before returning types without nil in https://github.com/ballerina-platform/ballerina-lang/blob/e9d7c054e4561ba41890e00bbe5240552f60664d/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L4823.

Reply:
Had a discussion with @gimantha and this should be a compilation error because `func()` is not lax.

Reply:
Let's add a test for this too.

Reply:
added